### PR TITLE
Filesystem limits

### LINKED ISSUE #3431: Filesystem size limits
Related to #3030 

My current plan is to provide two mechanisms for filesystem growth limits.

The first (and default) option will be limiting filesystems from growing past a certain size. This will be stored in the MDV. This limit can be changed or removed but can never be smaller than the current filesystem size.

The second option will be implementing XFS filesystem quotas. This aspect is a bit more complicated which is why I would like to keep it as an advanced option. Root users can bypass the quota per the design of XFS filesystem quotas which may lead to unexpected behavior so we would like users to opt in so that they understand the limitations.

I anticipate restricting filesystem size limits and XFS quotas from being used together. This will require tracking whether the limit is quota-based or size limit-based in the metadata.

Expected CLI invocation:
```
$ stratis filesystem limit-size 10GiB
$ stratis filesystem limit-size --quota 10GiB
Error
$ stratis filesystem limit-size 20GiB
$ stratis filesystem limit-size remove
$ stratis filesystem limit-size --quota 20GiB
$ stratis filesystem limit-size --quota remove
```

### PR DESCRIPTION
**Pull request by jbaublitz**
Closes #3431

### CODE REVIEW

**Comment by jbaublitz** [comment]
This is just waiting on unit tests.

**Comment by jbaublitz** [comment]
Just completed the unit tests.

**Review by bmr-cymru — CHANGES_REQUESTED** [review]

File: src/dbus_api/filesystem/shared.rs [hunk 1]
```diff
@@ -61,6 +63,116 @@ where
...
+{
+    i.append(
+        filesystem_get_operation(p.tree, p.path.get_name(), getter)
+            .map_err(|ref e| MethodErr::failed(e))?,
+    );
+    Ok(())
+}
+
+/// Set a filesystem property. The property is set by means of
+/// the getter method which takes a mutable reference to a
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/dbus_api/filesystem/shared.rs:86]: I think it should be setter here actually. We use setters within the closure.

[the remark this answers is not in the document]
Comment (bmr-cymru) [reply src/dbus_api/filesystem/shared.rs:86]: Yep, that makes sense to me!

**Review by mulkieran — CHANGES_REQUESTED** [review]
One minor request AND ought we to allow the user to specify the filesystem size limit on filesystem creation?

File: src/dbus_api/filesystem/mod.rs [hunk 2]
```diff
@@ -189,7 +191,8 @@ pub fn get_fs_properties(
             consts::FILESYSTEM_POOL_PROP => parent,
             consts::FILESYSTEM_CREATED_PROP => shared::fs_created_prop(fs),
             consts::FILESYSTEM_SIZE_PROP => shared::fs_size_prop(fs),
-            consts::FILESYSTEM_USED_PROP => shared::fs_used_prop(fs)
+            consts::FILESYSTEM_USED_PROP => shared::fs_used_prop(fs),
+            consts::FILESYSTEM_USED_PROP => shared::fs_size_limit_prop(fs)
```
Comment (mulkieran) [inline src/dbus_api/filesystem/mod.rs:195]: Plz fix duplicate entries here.

**Review by mulkieran — CHANGES_REQUESTED** [review]
Tests revealed a typo.

**Comment by jbaublitz** [comment]
@mulkieran I'm fine with letting the user specify filesystem size limits on creation if you'd like. It's more a matter of preference.

File: src/dbus_api/filesystem/mod.rs (hunk 2 above)
Comment (jbaublitz) [reply src/dbus_api/filesystem/mod.rs]: Resolved.

**Review by mulkieran — CHANGES_REQUESTED** [review]
When setting the filesystem size limit on creation, it's bounded by the maximum size of a thin device. I think that makes sense. But, I think then it also makes sense to enforce the same bound when setting the SizeLimit property.

**Review by mulkieran — CHANGES_REQUESTED** [review]
If it was possible to bring the sim engine into alignment w/ strat engine, so that setting the size limit property to below the current size returned an error, that would be helpful.

**Comment by jbaublitz** [comment]
@mulkieran Those two requests should now be resolved.

**Comment by mulkieran** [comment]
I want to put up the following proposal for a simplification in how we deal with the size limit property; we can talk about it in standup:

In stratisd:
1. Since we treat ```MAX_THIN_DEV_SIZE``` like a limit, treat it thus in ```StratFilesystem::extend_size```, i.e. explicitly cap the size at ```MAX_THIN_DEV_SIZE```. I expect that other constraints will cause that size never to be reached, but you never know.
2. Change the value representing the size limit in the engine so that it is non-optional.
3. Change the D-Bus property so that it non-optional and if no other limit has been set, report ```MAX_THIN_DEV_SIZE```.

In stratis-cli:
1. Always report the filesystem limit, as a number, in the filesystem listing under the heading: ```Total / Used / Free / Limit```.

The advantages of this are:
1. Some real simplification in the engine internals, due to non-optional value.
2. Some real simplification in the D-Bus API due to non-optional property.
3. Some real simplification in stratis-cli. This idea was prompted because of a request for an ```unset-size-limit``` command in stratis-cli. I would much rather keep the just the ```set-size-limit``` subcommand. It already takes one special value, ```stratis fs set-fs-limit <poolname> <fsname> current``` which just means set the filesystem limit to the filesystems current size, which seems sensible. I would prefer to add another special value for that subcommand, like ```stratis fs set-fs-limit <poolname> <fsname> largest```.  The way stratis would implement this request would be to have the constant ```MAX_THIN_DEV_SIZE``` defined; it would put that value in its property set call.

The disadvantages of this approach are:
1. User will see a very large limit in their ```filesystem list``` and may interpret it as meaning something different than it really does.

**Comment by drckeefe** [comment]
@mulkieran I agree that users might not know what the large value means.  Some might believe that this is a supported size maximum or a value to test, these might not be reasons to no do it.   I thought that Stratis might want to stay consistent with Key, where there is a set, unset, reset options, but the pool set-fs-limit doesn't have an unset or reset for good reasons.   Another thought based on the maximum sizes is,  does dm-thin and XFS share similar maximums?

**Comment by drckeefe** [comment]
@mulkieran Showing the size limit, do you have an example of what this might look like on the cli?  Imagining this in the output of engine_state_report seems clear.  I also wonder if there should be a stratis fs list command to view specific filesystems and their properties.  example: stratis fs list pool0 fs1

**Comment by mulkieran** [comment]
> @mulkieran I agree that users might not know what the large value means. Some might believe that this is a supported size maximum or a value to test, these might not be reasons to no do it. I thought that Stratis might want to stay consistent with Key, where there is a set, unset, reset options, but the pool set-fs-limit doesn't have an unset or reset for good reasons. Another thought based on the maximum sizes is, does dm-thin and XFS share similar maximums?

@drckeefe I don't think we should worry about staying consistent w/ key CLI commands, because setting and unsetting are fundamental actions for a key but setting and unsetting the fs size limit maybe not so fundamental for filesystems. That being said, there is nothing wrong w/ an ```unset-size-limit``` command, and we can implement the correct action via that command if that seems like it will be most intuitive to users. ```set-size-limit none``` is the alternative CLI command to do the same, and about as easy to implement, but might seem less consistent overall.

I think we just have to decide what "no limit" means, and once we settle that, any additional related CLI changes will be straightforward.

**Comment by mulkieran** [comment]
> @mulkieran Showing the size limit, do you have an example of what this might look like on the cli? Imagining this in the output of engine_state_report seems clear. I also wonder if there should be a stratis fs list command to view specific filesystems and their properties. example: stratis fs list pool0 fs1

@drckeefe For an example of what it looks like, see here: https://github.com/stratis-storage/stratis-cli/actions/runs/6528280013/job/17724252343?pr=1029#step:11:621. 

@drckeefe There should be such a command. There is an issue: https://github.com/stratis-storage/project/issues/510 .

**Comment by mulkieran** [comment]
I'm willing to believe that dropping ```MAX_THIN_DEV_SIZE``` entirely might be the truly simplifying move. But then I think we should drop it from the checks we do for the size value as well in order to be consistent w/ ourselves. In theory, then xfs will return an error on filesystem creation that stratisd will report. We usually try to let xfs do its own work, and I'm not sure why we didn't in this case.

> My objection to these changes is that in general:
> 
>     1. I think no filesystem limit is easy to understand. We can solicit feedback from @mvollmer on this if we'd like because I know he's had some opinions about usability in the past. I personally think it's less usable to display a very large number with no explanation and allow no way to remove it. That feels like leaking implementation details to the user in a way that's not helpful and will generate more questions than anything.
> 
>     2. I think that maintaining the `MAX_THINDEV_SIZE` in our codebase at all may be a mistake, but an even bigger mistake to mix it in with the filesystem limit code where it will be relatively hard to remove in the future. I don't like duplicating constants that cannot be set and may change in the future. Sometimes it's necessary, but I don't see enough of a benefit here to support that. Here I would almost rather have filesystem creation or extension fail so that we don't have to support multiple values across kernels if it changes in the future.
> 
>     3. I don't see the simplifications as large enough to justify the drawbacks.
> 
> 
> > I want to put up the following proposal for a simplification in how we deal with the size limit property; we can talk about it in standup:
> > In stratisd:
> > ```
> > 1. Since we treat `MAX_THIN_DEV_SIZE` like a limit, treat it thus in `StratFilesystem::extend_size`, i.e. explicitly cap the size at `MAX_THIN_DEV_SIZE`. I expect that other constraints will cause that size never to be reached, but you never know.
> > 
> > 2. Change the value representing the size limit in the engine so that it is non-optional.
> > 
> > 3. Change the D-Bus property so that it non-optional and if no other limit has been set, report `MAX_THIN_DEV_SIZE`.
> > ```
> > 
> > 
> > In stratis-cli:
> > ```
> > 1. Always report the filesystem limit, as a number, in the filesystem listing under the heading: `Total / Used / Free / Limit`.
> > ```
> > 
> > 
> > The advantages of this are:
> > ```
> > 1. Some real simplification in the engine internals, due to non-optional value.
> > 
> > 2. Some real simplification in the D-Bus API due to non-optional property.
> > 
> > 3. Some real simplification in stratis-cli. This idea was prompted because of a request for an `unset-size-limit` command in stratis-cli. I would much rather keep the just the `set-size-limit` subcommand. It already takes one special value, `stratis fs set-fs-limit <poolname> <fsname> current` which just means set the filesystem limit to the filesystems current size, which seems sensible. I would prefer to add another special value for that subcommand, like `stratis fs set-fs-limit <poolname> <fsname> largest`.  The way stratis would implement this request would be to have the constant `MAX_THIN_DEV_SIZE` defined; it would put that value in its property set call.
> > ```
> > 
> > 
> > The disadvantages of this approach are:
> > ```
> > 1. User will see a very large limit in their `filesystem list` and may interpret it as meaning something different than it really does.
> > ```

**Comment by jbaublitz** [comment]
Okay great! That's exactly what I wanted when I wrote the PR so I'll remove the `MAX_THINDEV_SIZE` altogether.

**Comment by jbaublitz** [comment]
@mulkieran Should we also remove `MIN_THINDEV_SIZE`?

**Comment by mulkieran** [comment]
> @mulkieran Should we also remove `MIN_THINDEV_SIZE`?

@jbaublitz The version of xfs in f38+ now enforces its lower bound so I think that we _might_ be able to do this. Let's discuss in standup.

**Comment by mulkieran** [comment]
whoops

**Review by mulkieran — CHANGES_REQUESTED** [review]
In the sim engine only, it looks like setting the size-limit on creation somehow doesn't cause the sim filesystem's size limit to be set. This is confirmed by ```stratis report``` which shows size limit as None for the sim engine and the set value is not displayed in the filesystem list view. For the strat engine, everything seems to be working correctly. For the sim engine and for the strat engine, setting and unsetting the size property seems to work correctly.

**Comment by jbaublitz** [comment]
> In the sim engine only, it looks like setting the size-limit on creation somehow doesn't cause the sim filesystem's size limit to be set. This is confirmed by `stratis report` which shows size limit as None for the sim engine and the set value is not displayed in the filesystem list view. For the strat engine, everything seems to be working correctly. For the sim engine and for the strat engine, setting and unsetting the size property seems to work correctly.

Resolved. This was caught by the unit tests for the strat_engine but isn't tested in the sim_engine so it wasn't caught.

**Comment by mulkieran** [comment]
@jbaublitz Plz rebase when you get the chance.

**Review by bgurney-rh — APPROVED** [review]

**Review by mulkieran — APPROVED** [review]

**Pull request merged**